arch/arm/mcount-insn.c

```diff
@@ -0,0 +1,146 @@
+#include "libmcount/internal.h"
+#include "mcount-arch.h"
+
+#define INSN_SIZE  8
+
+#ifdef HAVE_LIBCAPSTONE
+#include <capstone/capstone.h>
+#include <capstone/platform.h>
+
+void mcount_disasm_init(struct mcount_disasm_engine *disasm)
+{
+	if (cs_open(CS_ARCH_ARM, CS_MODE_ARM, &disasm->engine) != CS_ERR_OK) {
+		pr_dbg("failed to init Capstone disasm engine\n");
+		return;
+	}
+
+	if (cs_option(disasm->engine, CS_OPT_DETAIL, CS_OPT_ON) != CS_ERR_OK)
+		pr_dbg("failed to set detail option\n");
+}
+
+void mcount_disasm_finish(struct mcount_disasm_engine *disasm)
+{
+	cs_close(&disasm->engine);
+}
+
+/* return 0 if it's ok, -1 if not supported, 1 if modifiable */
+static int check_prologue(struct mcount_disasm_engine *disasm, cs_insn *insn)
+{
+	int i;
+	cs_arm *arm;
+	cs_detail *detail;
+	bool branch = false;
+	int status = -1;
+
+	/*
+	 * 'detail' can be NULL on "data" instruction
+	 * if SKIPDATA option is turned ON
+	 */
+	if (insn->detail == NULL)
+		return -1;
+
+	/* try to fix some PC-relative instructions */
+	if (insn->id == ARM_INS_ADR)
+		return 1;
+
+	/* check if the instruction is LDR, and it uses PC register */
+	if (insn->id == ARM_INS_LDR &&
+		 ((insn->bytes[2] & 0xf) == 0xf ||
+		  (insn->bytes[1] & 0xf0) == 0xf0)) {
```

Comment:
I guess there's better way to check it using capstone.

Reply:
Hmm.. I tried it as follows but capstone version is not matched. I will find another way.
```c
        cs_regs regs_read, regs_write;
        uint8_t regs_read_count, regs_write_count;

        /* check if the instruction uses PC register */
        if (cs_regs_access(disasm->engine, insn, regs_read, &regs_read_count,
                           regs_write, &regs_write_count)) {
                return -1;
        }

        for (i = 0; i < regs_read_count; i++) {
                if (regs_read[i] == ARM_REG_PC) {
                        fprintf(stderr, "read PC reg\n");
                        return -1;
                }
        }
        for (i = 0; i < regs_write_count; i++) {
                if (regs_write[i] == ARM_REG_PC) {
                        fprintf(stderr, "write PC reg\n");
                        return -1;
                }
        }
```

Reply:
Maybe it access PC implicitly.  What about checking base register of MEM type operand like x86?

```c
if (opnd2->type == ARM_OP_MEM && opnd2->mem.base == ARM_REG_PC)
   ...
```